File: solutions/718/main.py
from typing import List 

class Solution:
    def findLength(self, nums1: List[int], nums2: List[int]) -> int:
        result = 0
        # dp table
        dp = [[0] * len(nums2) for _ in range(len(nums1))]
        # initialize
        for i in range(len(nums1)):
            if nums1[i] == nums2[0]:
                dp[i][0] = 1
                result = 1
        for i in range(len(nums2)):
            if nums1[0] == nums2[i]:
                dp[0][i] = 1
                result = 1
        # traverse dp table
        for i in range(1, len(nums1)):
            for j in range(1, len(nums2)):
                if nums1[i] == nums2[j]:
                    dp[i][j] = dp[i-1][j-1] + 1
                if result < dp[i][j]:
                    result = dp[i][j]
        return result

if __name__ == "__main__":
    nums1 = [1,2,3,2,1]
    nums2 = [3,2,1,4,7]
    s = Solution()
    assert s.findLength(nums1, nums2) == 3

File: solutions/718/main.cpp
/* ************************************************************************
> File Name:     main.cpp
> Author:        Sean
> Project:    LeetCode
> Created Time:  Fri Nov 26 14:52:09 2021
> Description:   
 ************************************************************************/
#include <vector> 
#include <gtest/gtest.h> 

using namespace std; 

class Solution {
public:
    int findLength(vector<int>& nums1, vector<int>& nums2) {
        int result = 0;
        // dp table
        // dp[i][j] means the max len for nums1[0:i-1], and nums2[0:j-1]
        vector<vector<int>> dp(nums1.size()+1, vector<int>(nums2.size()+1, 0));
        // initialize
        // pass
        // traverse dp table
        for (int i = 1; i <= nums1.size(); ++i) {
            for (int j = 1; j <= nums2.size(); ++j) {
                if (nums1[i-1] == nums2[j-1]) {
                    dp[i][j] = dp[i-1][j-1] + 1;
                    if (result < dp[i][j]) result = dp[i][j];
                }
            }
        }
        return result;
    }
};

TEST(Test718, SimpleTest) {
    vector<int> nums1{
        1,2,3,2,1
    };
    vector<int> nums2{
        3,2,1,4,7
    };
    Solution s; 
    int result = s.findLength(nums1, nums2); 
    EXPECT_EQ(result, 3);
}

int main() {
    testing::InitGoogleTest(); 
    return RUN_ALL_TESTS();
}
